### ISSUE TITLE
Certificate error with "docker login"

### ISSUE THREAD

**Issue by rhasselbaum** [docker/docker#10150]
I'm setting up a private registry with a front-end Apache server serving as an SSL/TLS and Basic Auth proxy. I have a signed certificate for the Apache server. I placed the CA certificate in `/etc/docker/certs.d/<redacted>/ca.crt`.

All this works fine for `docker push` and `docker pull`. I see SSL connections being made. However, `docker login` fails with this error in the client:

```
FATA[0004] Error response from daemon: Server Error: Post https://<redacted>/v1/users/: x509: certificate signed by unknown authority
```

And I see the following in the daemon log:

```
DEBU[5219] Calling POST /auth                           
INFO[5219] POST /v1.16/auth                             
INFO[5219] +job auth()                                  
DEBU[5219] hostDir: /etc/docker/certs.d/<redacted> 
DEBU[5219] crt: /etc/docker/certs.d/<redacted>:443/ca.crt 
DEBU[5219] RegistryInfo.Version: ""                     
DEBU[5219] Registry standalone header: 'True'           
DEBU[5219] RegistryInfo.Standalone: true                
DEBU[5219] hostDir: /etc/docker/certs.d/<redacted> 
DEBU[5219] crt: /etc/docker/certs.d/<redacted>/ca.crt 
DEBU[5219] RegistryInfo.Version: ""                     
DEBU[5219] Registry standalone header: 'True'           
DEBU[5219] RegistryInfo.Standalone: true                
Server Error: Post https://<redacted>/v1/users/: x509: certificate signed by unknown authority
INFO[5219] -job auth() = ERR (1)                        
ERRO[5219] Handler for POST /auth returned error: Server Error: Post https://<redacted>/v1/users/: x509: certificate signed by unknown authority 
ERRO[5219] HTTP Error: statusCode=500 Server Error: Post https://<redacted>/v1/users/: x509: certificate signed by unknown authority 
```

I **know** it's reading the CA file because if I rename it, `docker pull` stops working. What would cause `docker pull` to work fine but `docker login` to fail with a certificate error? Here is my version info:

```
Client version: 1.4.1
Client API version: 1.16
Go version (client): go1.3.3
Git commit (client): 5bc2ff8
OS/Arch (client): linux/amd64
Server version: 1.4.1
Server API version: 1.16
Go version (server): go1.3.3
Git commit (server): 5bc2ff8
```

**Comment by rhasselbaum** [comment]
Installing the certificate globally works around the problem. Steps on Ubuntu/Debian:

1. Copy CA cert to `/usr/local/share/ca-certificates`.
3. sudo update-ca-certificates
4. sudo service docker restart

Once this is done, it doesn't matter if the cert is also loaded into `/etc/docker/certs.d/<registry>` or not.

**Comment by dillera** [comment]
Try this also:

Edit the DOCKER_OPTS variable
DOCKER_OPTS="-H unix:///var/run/docker.sock --insecure-registry your.repo.common.name"

**Comment by Thorndike** [comment]
Not working for me. Neither with port nor without makes any diffecerence. Still getting x509: certificate signed by unknown authority

**Comment by markllama** [comment]
From https://www.happyassassin.net/2014/09/06/adding-your-freeipa-servers-ca-certificate-to-the-system-wide-trust-store-on-fedora-and-rhel/

On RHEL you can add your docker registry server cert to the global CA list:

sudo cp coreos-reg.crt /etc/pki/ca-trust/source/anchors/
sudo update-ca-trust
sudo systemctl restart docker

**Comment by jfrazelle** [comment]
ping @dmcgowan @dmp42 can you help them resolve

**Comment by dmp42** [comment]
@jfrazelle smells like docker login codepath does not do the same thing as pull and fails to read `/etc/docker/certs.d/<redacted>/ca.crt`

Possibly @tiborvass can help pinpointing that (eg: the insecure thingie workflow)?

**Comment by jfrazelle** [comment]
closing as dup of https://github.com/docker/docker/issues/9570

**Comment by databus23** [comment]
@jfrazelle I don't think this issue is a duplicate of #9570. this issue ist about *docker login* not reading certificates from /etc/docker/... as *docker pull* does. The referenced issue is about sending credentials via plain http.

**Comment by jfrazelle** [comment]
i think the problem was also with @rhasselbaum setup because that should work if it was setup correctly, @rhasselbaum were you able to fix this

**Comment by databus23** [comment]
No this does not work and is a valid bug IMO.
@dmp42's assessment is correct. the code path of `docker login` does not add the certs from `/etc/docker/{host}` as `docker pull` does.
See:
  1. https://github.com/docker/docker/blob/master/registry/auth.go#L73
  2. https://github.com/docker/docker/blob/master/registry/registry.go#L36

**Comment by jfrazelle** [comment]
ok thanks for clarifying, reopening

**Comment by rhasselbaum** [comment]
I was only able to get certificate validation working with `docker login` by adding the certificate to the host's global CA trust store. Adding to `/etc/docker/certs.d/...` is still not working AFAIK.

@databus23 is correct, this is not the same issue as #9570.

**Comment by kateklemp** [comment]
I've added my cert to `/etc/docker/certs.d/<remote-reg>/ca.crt` and to `/usr/local/share/ca-certificates/ca.crt`.  I ran `sudo update-ca-certificates` and `sudo service docker restart`.  Docker login still gives me the same error:

```FATA[0005] Error response from daemon: v1 ping attempt failed with error: Get https://<remote-reg>/v1/_ping: x509: certificate signed by unknown authority. If this private registry supports only HTTP or HTTPS with an unknown CA certificate, please add `--insecure-registry <remote-reg>` to the daemon's arguments. In the case of HTTPS, if you have access to the registry's CA certificate, no need for the flag; simply place the CA certificate at /etc/docker/certs.d/<remote-reg>/ca.crt```

When I attempt to run with `--insecure registry <remote-reg>`, I get:

```flag provided but not defined: --insecure-registry
See 'docker login --help'.```

```
user@host:~ $ docker -v
Docker version 1.6.2, build 7c8fca2
user@host:~ $ docker login --help

Usage: docker login [OPTIONS] [SERVER]

Register or log in to a Docker registry server, if no server is
specified "https://index.docker.io/v1/" is the default.

  -e, --email=       Email
  --help=false       Print usage
  -p, --password=    Password
  -u, --username=    Username
```

Any update at all on this?

**Comment by dmcgowan** [comment]
@kateklemp the `--insecure-registry` flag needs to be set on the daemon, not the client.

For your certificate issue, are you able to curl the endpoint just fine using the ca certificate? 
```
$ curl --cacert=/etc/docker/certs.d/<remote-reg>/ca.crt https://<remote-reg>/v1/_ping
```

**Comment by screwgoth** [comment]
I have the above issue fixed for docker daemons running on Linux(Ubuntu).
However, when I run this using boot2linux for Mac, I get the "x509: certificate signed by unknown authority" error. 
I have copied the same ca.crt file at /etc/docker/certs.d/registry.example.com:8080 and restarted the dpcker daemon on the boot2docker VM, but still not able to solve this.

Anyone got this resolved ?

**Comment by dmp42** [comment]
@screwgoth have you tried with docker machine instead of boot2docker?

**Comment by winchell** [comment]
For those still having this problem having tried the above on Redhat / CentOS:

Try *manually* adding the chain certs to the end of /etc/pki/tls/certs/ca-bundle.crt. In my case running update-ca-trust didn't actualy update this file. I don't know the reason. There is some other formatting in ca-bundle.crt but don't worry about that and append the certs as they are in PEM format.

**Comment by ddysher** [comment]
I'm still experiencing this in docker 1.6, is there any fix going on?

My steup is pretty standard - an unsecure registry running behind nginx (ubuntu). I have to use the work around from @rhasselbaum.

Also, the problem doesn't seem to be just `docker login`, I can't do `docker pull/push` neither.

**Comment by bazsi** [comment]
I've also encountered this issue.

1) I am positive that I used the proper CA certificate (openssl s_client worked and reported successful validation)
2) I've installed that in /etc/certs.d/<myserver>/ca.crt
3) docker pull worked
4) docker login didn't

Once added into the root store, it worked like a charm. So definitely something is wrong wrt docker login in this case.

One note: when adding the cert into the root store, you need to add it to /etc/ssl/ca-certificates.crt and not just as a separate PEM file with a proper hashed symlink, docker seems to work only with the CA bundle and not the broken down CA certificate files).

Probably, if you are already logged in and your .dockercfg contains your credentials, it would work, but I guess any newcomers who wants to deploy their private registries will have a hard time figuring this out.

**Comment by Thorndike** [comment]
It seems --insecure-registry is working better now with docker 1.7.0

**Comment by bazsi** [comment]
I would not promote insecure-registry as a solution though. It is a workaround.

**Comment by omcnet** [comment]
Is there any effort to fix this bug or to unify code paths?

@bazsi You are absolutely right. Users must not disable certificate validation.
One of the main motivations for a private repository is a trusted data store for trustworthy downloads and execution.
If there is no guarantee, that the image is from one's own repository its like executing unknown code with superuser privileges...

**Comment by bbriggs** [comment]
Bump. Would like a solution here too, as I am managing a lot docker stuff via puppet and adding the cert to the global cert store with puppet will be a pain.

**Comment by seegler** [comment]
@markllama suggestion worked for me on Fedora-22
Steps with some minor modifications

openssl s_client -connect index.docker.io:443 -showcerts </dev/null 2>/dev/null | openssl x509 -outform PEM > docker.pem
sudo cp docker.pem /etc/pki/ca-trust/source/anchors/
sudo update-ca-trustsudo systemctl restart docker
sudo docker run hello-world

**Comment by mighq** [comment]
Is there any estimate where the real solution to this problem will be integrated?
I would like to see "login" command to look for CAs in the same location as "push/pull" commands (/etc/docker/certs.d/<repo>/ca.crt).

**Comment by jamiehannaford** [comment]
I'm also running into this error (`x509: certificate signed by unknown authority`). When I curl the endpoint it works, so I'd rather not have to fiddle with the Swarm host.

**Comment by grexe** [comment]
well for me it is the other way around, I can successfully login but not pull and get the same error mentioned by the original poster.
I tried all the solutions suggested but so far without success.
Using Docker 1.8.1 under Arch/Chakra x86 Linux:
```
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4
 Git commit:   d12ea79
 Built:        Die Sep  1 15:26:35 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4
 Git commit:   d12ea79
 Built:        Die Sep  1 15:26:35 UTC 2015
 OS/Arch:      linux/amd64
```

**Comment by chavenor** [comment]
@grexe  Did you ever get this figured out?

**Comment by spencerkohan** [comment]
@grexe where are you specifying the port exactly?  I mean, are you specifying that in the common name of the cert or what?

**Comment by grexe** [comment]
@spencerkohan no I was just referring to the directory which was created according to the registry name.
It is created according to the name given in the DOCKER_OPTS --insecure-registry URL.

Docker is not very consistent about this, e.g. for the local registry, you have to specify the port (which *has* to be 5000), but for a trusted (private) registry (which *has* to be SSL secured, god knows why...), you must *not* specify the port (443)...

**Comment by mishak87** [comment]
Fixed in 1.9.0

@rhasselbaum Was using 1.4.1 in 1.9.0 related code has been completely rewritten and as a result [http client setup unified](https://github.com/docker/docker/commit/e863a07b89599fd4a03d34491d67c09c6bc84444). As @databus23 pointed out http client setup was different between [`Login`](https://github.com/docker/docker/blob/v1.4.1/registry/auth.go#L172-L178) and rest of [`docker/docker/registry`](https://github.com/docker/docker/blob/v1.4.1/registry/registry.go#L38-L92) package.

[`Auth`](https://github.com/docker/docker/blob/249b3a9479048bcec8da0666c2e224dddcfc3de8/registry/service.go#L46) and `Pull` ([v1](https://github.com/docker/docker/blob/249b3a9479048bcec8da0666c2e224dddcfc3de8/registry/service_v1.go#L30) and [v2](https://github.com/docker/docker/blob/249b3a9479048bcec8da0666c2e224dddcfc3de8/registry/service_v2.go#L48)) share same [configuration code](https://github.com/docker/docker/blob/249b3a9479048bcec8da0666c2e224dddcfc3de8/registry/endpoint.go#L48). And [`TLSConfig`](https://github.com/docker/docker/blob/249b3a9479048bcec8da0666c2e224dddcfc3de8/registry/service.go#L111) is just a helper.

Issue what @grexe is that docker login and pull accept `https://something:443`, `https://something:443` and `something` and see it as 3 different registries instead of one which can be confusing for users.

@grexe I think your issue is with mixup of different login and pull urls please correct me if I'm wrong.

**Comment by thaJeztah** [comment]
@rhasselbaum is this issue resolved for you in docker 1.9? (following the discussion above)

**Comment by tallandtree** [comment]
For anyone who still has this problem with docker 1.9. I also had this error (x509 certificate error) and finally found the cause: We are behind a corporate http proxy, so I configured a proxy in /etc/systemd/system/docker.service.d/http-proxy.conf but forgot to add my new private registry server to the NO_PROXY list. Curl respects the local exported "no_proxy" setting, but docker needs to know this as well via the above configuration file.  Error message was a tad confusing.

**Comment by ecirtap** [comment]
@tallandtree thanks a lot ! Same thing here (corporate proxy)

curl ok
docker login fail
http_proxy was set
no_proxy was not set

**Comment by Akanksha08** [comment]
Is this still a valid bug? If so I would like to work on this.

**Comment by JMGordon** [comment]
@Akanksha08 reply to the issue with "#dibs".

**Comment by JMGordon** [comment]
@Akanksha08 just reply to the issue with "dibs" preceded with a "#". Don't mind the "status/claimed" label. I wrote a message to explain the previous statement and accidentally claimed the issue.

**Comment by Akanksha08** [comment]
#dibs

**Comment by smontosa** [comment]
I'm using a certificate from https://letsencrypt.org/ . I'm sure there is any official way to make that docker accepts its CA certificate but I did it on the "wild" way. I concatenated ca.pem from https://letsencrypt.org/certificates/ (https://letsencrypt.org/certs/lets-encrypt-x3-cross-signed.pem for me) to cert.pem file at certs/ folder at registry's volume. That worked for me.

cat ca.pem mycert.perm > cert.pem

Here, part of my docker-compose file:

environment:
    REGISTRY_HTTP_TLS_CERTIFICATE: /certs/cert.pem
    REGISTRY_HTTP_TLS_KEY: /certs/privkey.pem
    REGISTRY_AUTH: htpasswd
    REGISTRY_AUTH_HTPASSWD_PATH: /auth/htpasswd
    REGISTRY_AUTH_HTPASSWD_REALM: Registry Realm
  volumes:
    - ./data:/var/lib/registry
    - ./certs:/certs
    - ./auth:/auth

**Comment by dmcgowan** [comment]
@smontosa the CA needs to be trusted on the host the docker daemon is running on. This can either be done by adding to the system chain (which Let's Encrypt is widely trusted already), or to `/etc/docker/certs.d/{registryhost}`.

**Comment by smontosa** [comment]
@dmcgowan Thank you for your explanation. Maybe my docker installation is old or  it has became to be a bit messy. I'll try it again and follow your advice.

**Comment by ghostplant** [comment]
Hi all, I encountered this issue as well and I finally figure out the reason and solution.
If the Docker Registry is using a pair of certs.{key,pem} that is generated by a newly version of OpenSSL, while Docker Client is using an old-version OpenSSL library, this error might affect you.
So a general solution is to upgrade your client-side OpenSSL library to the newest version, for example, >=1.0.2

**Comment by lachnerd** [comment]
Hi,
the comment from "winchell commented on 31 May 2015" made my day.

I had a fresh official wildcard certificate but my artifactory with nginx in front still complaint about:
`v1/users/: x509: certificate signed by unknown authority`

So i manually added the chain certificate at the end of the cert.crt restartet nginx and boom it worked finally.

Maybe this helps someone else

**Comment by chankongching** [comment]
@lachnerd Your way works like a charm for me, thanks.

**Comment by vikash315** [comment]
Hi,

I have installed docker in windows 7 using Windows Tool Box. I getting the same error while login to IBM bluemix registry. 

D:\Project\LH\source_code\crew-safety-service-api>docker login registry.ng.bluemix.net
Username: vikash315@gmail.com
Password:
Error response from daemon: Get https://registry.ng.bluemix.net/v1/users/: x509: certificate signed by unknown authority 

I have read above comments but didn't find any clue to fix it in windows.

Please suggests How to fix this issue.

Thanks,
Vikash

**Comment by friism** [comment]
@vikash315 Are you behind a proxy or similar? Please see:

 * https://github.com/docker/toolbox/issues/603#issuecomment-266020709
 * http://stackoverflow.com/a/36454369/2942

**Comment by usptact** [comment]
@tallandtree So much this! I literally wasted one day trying different docker versions, messing with system/docker certificates, various configuration options... just to discover my no_proxy was not properly set (there was an extra " symbol which made it fail parsing the value).
